runtime/server/psql.go

```diff
@@ -0,0 +1,252 @@
+		}
+		return wire.Prepared(wire.NewStatement(handle, wire.WithColumns(convertSchema(schema)))), nil
+	}
+
+	// todo handle normal SQL. Some ideas to handle that
+	// 1. Approach similar to BigQuery which has #legacySQL
+	// 2. We can parse for a -- @connector: xxx comment, and if found, use that connector directly instead of metrics SQL
+	res, err := s.runtime.Resolve(ctx, &runtime.ResolveOptions{
+		InstanceID:         instanceID,
+		Resolver:           "metrics_sql",
+		ResolverProperties: map[string]any{"sql": query},
+		Args:               map[string]any{"priority": 1},
+		UserAttributes:     auth.GetClaims(ctx).Attributes(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return wire.Prepared(wire.NewStatement(s.execQuery(instanceID, query), wire.WithColumns(convertSchema(res.Schema)))), nil
+}
+
+func (s *Server) execQuery(instanceID, query string) wire.PreparedStatementFn {
+	return func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
+		// NOTE :: the duplicate call to runtime.Resolve is made so that we do not keep the results and eventually leak memory if query is never executed.
+		// This should resolve fast in most cases since results will be cached.
+		res, err := s.runtime.Resolve(ctx, &runtime.ResolveOptions{
+			InstanceID:         instanceID,
+			Resolver:           "metrics_sql",
+			ResolverProperties: map[string]any{"sql": query},
+			Args:               map[string]any{"priority": 1},
+			UserAttributes:     auth.GetClaims(ctx).Attributes(),
+		})
+		if err != nil {
+			return err
+		}
+
+		var rawData []map[string]interface{}
+		err = json.Unmarshal(res.Data, &rawData)
+		if err != nil {
+			return err
+		}
+
+		row := make([]any, len(res.Schema.Fields))
+		for i := 0; i < len(rawData); i++ {
+			// get the values in the same order as schema
+			for j := 0; j < len(res.Schema.Fields); j++ {
+				if v, ok := rawData[i][res.Schema.Fields[j].Name]; ok {
+					// do type conversion if required
+					row[j], err = convertValue(v, res.Schema.Fields[j].Type.Code)
+					if err != nil {
+						return err
+					}
+				} else {
+					row[j] = nil
+				}
+			}
+			if err := writer.Row(row); err != nil {
+				return fmt.Errorf("failed to write row: %w", err)
+			}
+		}
+		return writer.Complete("OK")
+	}
+}
+
+func convertValue(v any, code runtimev1.Type_Code) (any, error) {
+	if v == nil {
+		return nil, nil
+	}
+	switch code {
+	case runtimev1.Type_CODE_DATE:
+		if u, ok := v.(string); ok {
+			return time.Parse(time.DateOnly, u)
+		}
+		return v, nil
+	case runtimev1.Type_CODE_TIMESTAMP:
+		if u, ok := v.(string); ok {
+			return time.Parse(time.RFC3339, u)
+		}
+		return v, nil
+	default:
+		return v, nil
+	}
+}
+
+func convertSchema(schema *runtimev1.StructType) wire.Columns {
+	columns := make([]wire.Column, len(schema.Fields))
+	for i, field := range schema.Fields {
+		columns[i] = columnForType(field)
+	}
+	return columns
+}
+
+func columnForType(field *runtimev1.StructType_Field) wire.Column {
```

Comment:
Same comment about function name specificity here – would be fine in a small package, but in the context of `runtime/server`, these names could mean a lot of different things.

Reply:
Sure. I think in that case will add these as methods of a struct.